### PR TITLE
FFT based M2L

### PR DESCRIPTION
**Pull request by isuruf**
Replaces https://github.com/inducer/sumpy/pull/36

For FFT based M2L, we first need to take the FFT of the source coeffs. This is given by
`sumpy.expansion.local.LocalExpansionBase.m2l_preprocess_exprs` and done
separately in a new loopy kernel produced by `sumpy.e2e.E2EFromCSRWithFFTPreprocess`.
This happens in a separate loopy kernel because the FFT of the inputs needs to happen
only once for each source box. Since the M2L is done for each target box, if we didn't do the
FFT separately, source boxes will be processed using FFT multiple times.

After that the derivatives which are precomputed for translation classes need to be
FFTed as well. This is done in the same loopy kernel that computes the derivatives.

M2L is then just the elementwise multiplication of the FFT of the source coeffs and
the FFT of the derivatives. After adding these for each target box, we do an inverse FFT
for each target box. This needs to happen outside the inner loop of iterating the source
boxes in M2L to avoid redundant calculations. The expressions involved are given by
`sumpy.expansion.local.LocalExpansionBase.m2l_postprocess_exprs`.

One other major change is that the loopy instructions depend on whether the input to
M2L is real or not. If it's real, FFT and then inverse FFT gives us complex expressions
with imaginary part close to zero. We need to truncate the imaginary part if the input was
real, but keep the imaginary part as it is if the input was complex. Therefore
`E2EFromCSE.get_kernel` gained an additional parameter `result_dtype`.

### CODE REVIEW

**Comment by inducer** [comment]
(Did you mean to leave this on draft? Just checking)

**Comment by isuruf** [comment]
Yes, will let you know when it's ready for review.

**Comment by isuruf** [comment]
Ready for a review now. Added a description in the PR body which give an overview.

File: sumpy/fmm.py [hunk 1]
```diff
@@ -415,6 +440,31 @@ def m2l_precomputed_exprs_view(self, m2l_precomputed_exprs, level):
         return (translation_class_start, exprs_level.reshape(
                             translation_class_stop - translation_class_start, -1))
 
+    @memoize_method
+    def pp_multipole_expansions_level_starts(self):
```
Comment (inducer) [inline sumpy/fmm.py:444]:
```suggestion
    def m2l_preproc_mpole_expansions_level_starts(self):
``` 
(also below)

File: sumpy/fmm.py [hunk 2]
```diff
@@ -639,10 +689,42 @@ def multipole_to_local(self,
...
+                source_level_start_ibox, source_mpoles_view = \
+                        self.multipole_expansions_view(mpole_exps, lev)
+
+                _, pp_source_mpoles_view = \
+                        self.pp_multipole_expansions_view(pp_mpole_exps, lev)
+
+                tr_classes = self.m2l_translation_class_level_start_box_nrs()
+                # There's no M2L happening in this level
+                if tr_classes[lev] == tr_classes[lev + 1]:
+                    continue
```
Comment (inducer) [inline sumpy/fmm.py:710]:
```suggestion
                if tr_classes[lev] == tr_classes[lev + 1]:
                    # There's no M2L happening in this level
                    continue
```

File: sumpy/fmm.py [hunk 3]
```diff
@@ -639,10 +689,42 @@ def multipole_to_local(self,
...
+                )
+                events.append(evt)
+            mpole_exps = pp_mpole_exps
+            mpole_exps_view_func = self.pp_multipole_expansions_view
+        else:
+            mpole_exps_view_func = self.multipole_expansions_view
+
         local_exps = self.local_expansion_zeros()
 
         events = []
```
Comment (inducer) [inline sumpy/fmm.py:727]: *cough* event management

File: sumpy/symbolic.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ def _find_symbolic_backend():
 # Before adding a function here, make sure it's present in both modules.
 SYMBOLIC_API = """
 Add Basic Mul Pow exp sqrt log symbols sympify cos sin atan2 Function Symbol
-Derivative Integer Matrix Subs I pi functions Number""".split()
+Derivative Integer Matrix Subs I pi functions Number Float""".split()
```
Comment (inducer) [inline sumpy/symbolic.py:74]: ?

File: sumpy/tools.py [hunk 5]
```diff
@@ -862,4 +911,18 @@ def matvec_toeplitz_upper_triangular(first_row, vector):
...
+def to_complex_dtype(dtype):
+    if dtype == np.complex64:
+        return np.complex64
+    if dtype == np.complex128:
+        return np.complex128
+    if dtype == np.float32:
+        return np.complex64
+    if dtype == np.float64:
+        return np.complex128
+    raise RuntimeError(f"Unknown dtype: {dtype}")
```
Comment (inducer) [inline sumpy/tools.py:924]: `dict`?

File: test/test_fmm.py [hunk 6]
```diff
@@ -51,36 +51,46 @@
...
+        (HelmholtzKernel(2), H2DLocalExpansion,
+            H2DMultipoleExpansion, True, False),
+        #(HelmholtzKernel(2), H2DLocalExpansion,
+        #    H2DMultipoleExpansion, True, True),
+        (HelmholtzKernel(3), VolumeTaylorLocalExpansion,
+            VolumeTaylorMultipoleExpansion, False, False),
+        (HelmholtzKernel(3), LinearPDEConformingVolumeTaylorLocalExpansion,
+            LinearPDEConformingVolumeTaylorMultipoleExpansion, False, False),
+        (YukawaKernel(2), Y2DLocalExpansion,
+            Y2DMultipoleExpansion, False, False),
```
Comment (inducer) [inline test/test_fmm.py:90]: *cough* loops *cough*?

File: sumpy/e2e.py [hunk 7]
```diff
@@ -190,19 +200,84 @@ def get_translation_loopy_insns(self):
...
                 self.tgt_expansion.m2l_global_precompute_nexpr(self.src_expansion)
         else:
             nprecomputed_exprs = 0
 
+        if self.use_fft:
+            ncoeff_src = nprecomputed_exprs
+            ncoeff_post = nprecomputed_exprs
+        else:
+            ncoeff_src = len(self.src_expansion)
+            ncoeff_post = len(self.tgt_expansion)
```
Comment (inducer) [inline sumpy/e2e.py:277]: ?

File: sumpy/e2e.py [hunk 8]
```diff
@@ -424,21 +517,120 @@ def __call__(self, queue, **kwargs):
...
+                vector_names=set(["d"]),
+                pymbolic_expr_maps=[self.tgt_expansion.get_code_transformer()],
+                retain_names=pp_src_coeff_names,
+                complex_dtype=to_complex_dtype(result_dtype),
+                )
+
+    def get_kernel(self, result_dtype):
+        nsrc_coeffs = len(self.src_expansion)
+        npp_src_coeffs = \
+            self.tgt_expansion.m2l_global_precompute_nexpr(self.src_expansion)
```
Comment (inducer) [inline sumpy/e2e.py:577]: ?

File: sumpy/e2e.py [hunk 9]
```diff
@@ -424,21 +517,120 @@ def __call__(self, queue, **kwargs):
...
                 m2l_translation_vectors=m2l_translation_vectors,
+                m2l_precomputed_exprs=m2l_precomputed_exprs,
                 **kwargs)
 
 # }}}
 
 
+# {{{ class for calculating the FFT of multipole expansion for M2L
+
+class E2EFromCSRWithFFTPreprocess(E2EFromCSR):
```
Comment (inducer) [inline sumpy/e2e.py:541]: * Shouldn't inherit from CSR.
* Naming: no FFT

File: sumpy/e2e.py [hunk 10]
```diff
@@ -424,21 +517,120 @@ def __call__(self, queue, **kwargs):
...
+    """Implements preprocessing the multipole expansions
+    for M2L with FFT.
```
Comment (inducer) [inline sumpy/e2e.py:543]:
```suggestion
    """Computes the preprocessed multipoled coefficients for accelerated
    M2L.
```

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: sumpy/symbolic.py (hunk 4 above)
Comment (isuruf) [reply sumpy/symbolic.py:74]: This is used in a test now.

File: sumpy/expansion/local.py [hunk 11]
```diff
@@ -47,10 +48,37 @@
...
+    .. attribute:: use_rscale
+    .. attribute:: use_fft
+
     .. automethod:: m2l_global_precompute_exprs
     .. automethod:: m2l_global_precompute_nexpr
     .. automethod:: translate_from
     """
+    init_arg_names = ("kernel", "order", "use_rscale", "use_fft")
+
+    def __init__(self, kernel, order, use_rscale=None, use_fft=False):
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 25cfae7bbf5ea21fc6f98c682cbcafe89ae76524

File: sumpy/expansion/local.py [hunk 12]
```diff
@@ -76,6 +104,28 @@ def m2l_global_precompute_nexpr(self, src_expansion):
         """
         return 0
 
+    def m2l_preprocess_exprs(self, src_expansion, src_coeff_exprs, sac,
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 9c64d12466e87af11f9881e0ed55f9d02d1659c8

File: sumpy/expansion/local.py [hunk 13]
```diff
@@ -76,6 +104,28 @@ def m2l_global_precompute_nexpr(self, src_expansion):
...
+            src_rscale):
+        """Return preprocessed input expressions for the Toeplitz matrix
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 4add1a6e3b0a5be790c7a5112872aa9900afeaf1

* Precompute -> m2l_translation_classes_dependent_data
* preprocess -> m2l_preprocess_multipole_exprs
* postprocess -> m2l_postprocess_local_exprs

File: sumpy/expansion/local.py [hunk 14]
```diff
@@ -47,10 +48,37 @@
...
+
     .. automethod:: m2l_global_precompute_exprs
     .. automethod:: m2l_global_precompute_nexpr
     .. automethod:: translate_from
     """
+    init_arg_names = ("kernel", "order", "use_rscale", "use_fft")
+
+    def __init__(self, kernel, order, use_rscale=None, use_fft=False):
+        super().__init__(kernel, order, use_rscale)
+        self.use_fft = use_fft
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 25cfae7bbf5ea21fc6f98c682cbcafe89ae76524

File: sumpy/expansion/local.py [hunk 15]
```diff
@@ -227,7 +277,10 @@ def evaluate(self, kernel, coeffs, bvec, rscale, sac=None):
     def m2l_global_precompute_nexpr(self, src_expansion):
         """Returns number of expressions in M2L global precomputation step.
         """
-        return len(self._m2l_global_precompute_mis(src_expansion)[1])
+        if self.use_fft:
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 361d0fa26ecc86579b78cf2694f9fdf181d8d4fd

File: sumpy/expansion/local.py [hunk 16]
```diff
@@ -339,8 +396,57 @@ def m2l_global_precompute_exprs(self, src_expansion, src_rscale,
         for i, term in enumerate(needed_vector_terms):
             vector[i] = add_to_sac(sac,
                         vector_full[srcplusderiv_ident_to_index[term]])
+
+        if self.use_fft:
+            # Add zero values needed to make the translation matrix toeplitz
+            derivatives_full = [0]*len(toeplitz_matrix_coeffs)
+            for expr, mi in zip(vector, needed_vector_terms):
+                derivatives_full[toeplitz_matrix_ident_to_index[mi]] = expr
+            return fft(list(reversed(derivatives_full)), sac=sac)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py:445]: Done in 535db51c4b988daa4e1591857675b97bec5594de

File: sumpy/expansion/local.py [hunk 17]
```diff
@@ -339,8 +396,57 @@ def m2l_global_precompute_exprs(self, src_expansion, src_rscale,
         for i, term in enumerate(needed_vector_terms):
             vector[i] = add_to_sac(sac,
                         vector_full[srcplusderiv_ident_to_index[term]])
+
+        if self.use_fft:
+            # Add zero values needed to make the translation matrix toeplitz
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 535db51c4b988daa4e1591857675b97bec5594de

File: sumpy/expansion/local.py [hunk 18]
```diff
@@ -339,8 +396,57 @@ def m2l_global_precompute_exprs(self, src_expansion, src_rscale,
...
+        # Calculate the first row of the upper triangular Toeplitz matrix
+        toeplitz_first_row = [0] * len(toeplitz_matrix_coeffs)
+        for coeff, term in zip(
+                src_coeff_exprs,
+                src_expansion.get_coefficient_identifiers()):
+            toeplitz_first_row[toeplitz_matrix_ident_to_index[term]] = \
+                    add_to_sac(sac, coeff)
+
+        if self.use_fft:
+            return fft(toeplitz_first_row, sac=sac)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Fixed in 535db51c4b988daa4e1591857675b97bec5594de

File: sumpy/expansion/local.py [hunk 19]
```diff
@@ -339,8 +396,57 @@ def m2l_global_precompute_exprs(self, src_expansion, src_rscale,
...
+        else:
+            return toeplitz_first_row
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 0a4f12732297a21840729799bbf9fb1f40452097

File: sumpy/expansion/local.py [hunk 20]
```diff
@@ -339,8 +396,57 @@ def m2l_global_precompute_exprs(self, src_expansion, src_rscale,
...
+            n = len(toeplitz_matrix_coeffs)
+            m2l_result = fft(m2l_result, inverse=True, sac=sac)
+            m2l_result = list(reversed(m2l_result[:n]))
+
+        # Filter out the dummy rows and scale them for target
+        result = []
+        rscale_ratio = add_to_sac(sac, tgt_rscale/src_rscale)
+        for term in self.get_coefficient_identifiers():
+            index = toeplitz_matrix_ident_to_index[term]
+            result.append(m2l_result[index]*rscale_ratio**sum(term))
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 818ab9cdef313bf869930f9f0f3f57d77432f67b

File: sumpy/expansion/local.py [hunk 21]
```diff
@@ -371,28 +477,33 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
 
+            if self.use_fft:
+                assert precomputed_exprs is not None
+                assert len(src_coeff_exprs) == len(precomputed_exprs)
+                result = []
+                for i in range(len(precomputed_exprs)):
+                    a = precomputed_exprs[i]
+                    b = src_coeff_exprs[i]
+                    result.append(a*b)
+                return result
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 0677208f3a8004df588f04c77f71a16935e7c6bf

File: sumpy/expansion/local.py [hunk 22]
```diff
@@ -371,28 +477,33 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
-                    src_expansion.get_coefficient_identifiers()):
-                toeplitz_first_row[toeplitz_matrix_ident_to_index[term]] = \
-                        add_to_sac(sac, coeff)
+            toeplitz_first_row = self.m2l_preprocess_exprs(src_expansion,
+                src_coeff_exprs, sac, src_rscale)
 
             # Do the matvec
-            output = matvec_toeplitz_upper_triangular(toeplitz_first_row,
+            if 0:
+                output = fft_toeplitz_upper_triangular(toeplitz_first_row,
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Removed the debug code path

File: sumpy/expansion/local.py [hunk 23]
```diff
@@ -630,9 +743,69 @@ def evaluate(self, kernel, coeffs, bvec, rscale, sac=None):
...
+    def m2l_global_precompute_exprs(self, src_expansion, src_rscale,
+            dvec, tgt_rscale, sac):
+
+        from sumpy.symbolic import sym_real_norm_2, Hankel1
+        from sumpy.tools import fft
+
+        dvec_len = sym_real_norm_2(dvec)
+        new_center_angle_rel_old_center = sym.atan2(dvec[1], dvec[0])
+        arg_scale = self.get_bessel_arg_scaling()
+        precomputed_exprs = [0] * (2*self.order + 2 * src_expansion.order + 1)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Done in 84bdc2011337e51116c61b9c7b742baecf6160e5

File: sumpy/expansion/local.py [hunk 24]
```diff
@@ -630,9 +743,69 @@ def evaluate(self, kernel, coeffs, bvec, rscale, sac=None):
...
+        for j in self.get_coefficient_identifiers():
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py:808]: Done in 84bdc201

File: sumpy/expansion/local.py [hunk 25]
```diff
@@ -676,12 +862,13 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
 class H2DLocalExpansion(_FourierBesselLocalExpansion):
-    def __init__(self, kernel, order, use_rscale=None):
+    def __init__(self, kernel, order, use_rscale=None, use_fft=False):
         from sumpy.kernel import HelmholtzKernel
         assert (isinstance(kernel.get_base_kernel(), HelmholtzKernel)
                 and kernel.dim == 2)
 
-        super().__init__(kernel, order, use_rscale)
+        super().__init__(kernel, order, use_rscale, use_fft=use_fft)
+        print(use_fft)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Removed.

File: sumpy/fmm.py [hunk 26]
```diff
@@ -293,14 +302,23 @@ def __init__(self, code_container, queue, tree, dtype, fmm_level_to_order,
...
 
         self.dtype = dtype
 
+        if not self.code.use_fft:
+            # If not FFT, we don't need complex dtypes
+            self.complex_dtype = dtype
+        elif complex_dtype is not None:
+            self.complex_dtype = complex_dtype
+        else:
+            self.complex_dtype = to_complex_dtype(dtype)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/fmm.py]: Changed to `preprocessed_mpole_dtype` in c12f176a569e1d64d68f5c876e74a5cc5ba430bd

File: sumpy/fmm.py (hunk 1 above)
Comment (isuruf) [reply sumpy/fmm.py]: Done in 5d32bfbe4853ea92e572e8c46f185e6084317dd5

File: sumpy/fmm.py (hunk 2 above)
Comment (isuruf) [reply sumpy/fmm.py]: Done in a7d4055e33606c3c18bc93eaa7b83704b6431456

File: sumpy/fmm.py (hunk 3 above)
Comment (isuruf) [reply sumpy/fmm.py]: Done in 467e43b6bd8fac000e4c8f0cafc2cdf76f85122a

File: sumpy/tools.py (hunk 5 above)
Comment (isuruf) [reply sumpy/tools.py]: Fixed in 2d06371313e990320852024477a70e9c535c6142

File: test/test_fmm.py (hunk 6 above)
Comment (isuruf) [reply test/test_fmm.py]: Fixed in 22e493063d70be8726dc72c41e855989b6c46789

File: sumpy/e2e.py (hunk 7 above)
Comment (isuruf) [reply sumpy/e2e.py]: Clarified in d83705cb388fe230eff6801220416df36b63bdb0

File: sumpy/e2e.py (hunk 8 above)
Comment (isuruf) [reply sumpy/e2e.py]: changed in the refactor at 4add1a6e

File: sumpy/e2e.py (hunk 9 above)
Comment (isuruf) [reply sumpy/e2e.py]: Changed to `M2LPreprocessMultipole` in 4add1a6e

File: sumpy/e2e.py (hunk 10 above)
Comment (isuruf) [reply sumpy/e2e.py]: Fixed in 8045c01db3af4f63f7aa4bb19ed95998f79e304b

**Comment by isuruf** [comment]
@inducer, this is ready.

Note that I have disabled FFT tests for H2DLocal and Y2DLocal because they are inaccurate. The scaling here is not working properly. I'll try to fix that in a follow-up PR if that's okay.

File: sumpy/fmm.py (hunk 3 above)
Comment (inducer) [reply sumpy/fmm.py]: This isn't right yet. I'll fix it up.

File: sumpy/expansion/local.py [hunk 27]
```diff
@@ -676,12 +926,17 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
         from sumpy.kernel import HelmholtzKernel
         assert (isinstance(kernel.get_base_kernel(), HelmholtzKernel)
                 and kernel.dim == 2)
 
-        super().__init__(kernel, order, use_rscale)
+        super().__init__(kernel, order, use_rscale,
+                use_preprocessing_for_m2l=use_preprocessing_for_m2l)
+
+        if use_preprocessing_for_m2l:
+            raise ValueError("H2DLocalExpansion with FFT is not implemented yet.")
```
Comment (inducer) [inline sumpy/expansion/local.py:939]: * Comment on the state here. (You mention that sympy certifies equality, but numerically "something is up". Also, potential `rscale` issues.)
* Move exception to base class?

File: sumpy/fmm.py [hunk 28]
```diff
@@ -287,20 +299,33 @@ class SumpyExpansionWrangler:
...
 
         self.dtype = dtype
 
+        if not self.code.use_preprocessing_for_m2l:
+            # If not FFT, we don't need complex dtypes
+            self.preprocessed_mpole_dtype = dtype
+        elif preprocessed_mpole_dtype is not None:
+            self.preprocessed_mpole_dtype = preprocessed_mpole_dtype
+        else:
+            self.preprocessed_mpole_dtype = to_complex_dtype(dtype)
```
Comment (inducer) [inline sumpy/fmm.py:327]: It's weird that the wrangler has to guess this. (If this doesn't want to go easily, add a FIXME comment.)

File: sumpy/tools.py [hunk 29]
```diff
@@ -862,4 +911,21 @@ def matvec_toeplitz_upper_triangular(first_row, vector):
...
+to_complex_dtype_dict = {
+    np.complex64: np.complex64,
+    np.complex128: np.complex128,
+    np.float32: np.complex64,
+    np.float64: np.complex128,
+}
+
+
+def to_complex_dtype(dtype):
+    try:
```
Comment (inducer) [inline sumpy/tools.py:924]:
```suggestion
    dtype = np.dtype(dtype).type
    try:
```

**Review by inducer — COMMENTED** [review]
:tada: This is getting close!

File: sumpy/e2e.py [hunk 30]
```diff
@@ -84,6 +84,11 @@ def __init__(self, ctx, src_expansion, tgt_expansion,
         self.tgt_expansion = tgt_expansion
         self.name = name or self.default_name
         self.device = device
+        try:
+            self.use_preprocessing_for_m2l \
+                = self.tgt_expansion.use_preprocessing_for_m2l
+        except AttributeError:
+            self.use_preprocessing_for_m2l = False
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/e2e.py]: Fixed in 452180fe7e2d6c274ead26e8d0d9686a6b9c5ce2

File: sumpy/e2e.py [hunk 31]
```diff
@@ -148,30 +153,42 @@ class E2EFromCSR(E2EBase):
...
         from sumpy.symbolic import make_sym_vector
         dvec = make_sym_vector("d", self.dim)
 
-        src_coeff_exprs = [sym.Symbol("src_coeff%d" % i)
-                for i in range(len(self.src_expansion))]
         src_rscale = sym.Symbol("src_rscale")
 
         tgt_rscale = sym.Symbol("tgt_rscale")
 
         extra_kwargs = dict()
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/e2e.py]: Fixed in 9045aa69558fe9557353ebee30f6ae43743c005e

File: sumpy/e2e.py [hunk 32]
```diff
@@ -421,24 +537,128 @@ def __call__(self, queue, **kwargs):
...
+        loopy_knl = lp.make_kernel(
+                [
+                    "{[isrc_box]: 0<=isrc_box<nsrc_boxes}",
+                    ],
+                ["""
+                for isrc_box
+                """] + ["""
+                    <> src_coeff{idx} = src_expansions[isrc_box, {idx}]
+                """.format(idx=i) for i in range(nsrc_coeffs)] + [
+                ] + self.get_translation_loopy_insns(result_dtype) + ["""
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/e2e.py]: Fixed in b12efdc529e5f5811cba6c7cd7adb2498f6d7a7c

File: sumpy/expansion/local.py [hunk 33]
```diff
@@ -261,17 +348,17 @@ def _m2l_global_precompute_mis(self, src_expansion):
                               self.get_coefficient_identifiers())
 
         # These are the multi-indices representing the rows
-        # in the Toeplitz matrix.  Note that to get the Toeplitz
+        # in the circulant matrix.  Note that to get the circulant
         # matrix structure some multi-indices that is not in the
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Fixed in 6ed5f6bb0755622457e48870b023ec6d0a9c63fd

File: sumpy/expansion/local.py [hunk 34]
```diff
@@ -261,17 +348,17 @@ def _m2l_global_precompute_mis(self, src_expansion):
...
         # M2L translation matrix are added.
         # This corresponds to adding $\mathcal{O}(p^{d-1})$
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Fixed in fb662f0f4080897c8cbb7e3256492ffd24f05f7a

File: sumpy/expansion/local.py [hunk 35]
```diff
@@ -303,8 +391,11 @@ def m2l_global_precompute_exprs(self, src_expansion, src_rscale,
         if not self.use_rscale:
             src_rscale = 1
 
-        toeplitz_matrix_coeffs, needed_vector_terms, max_mi = \
-            self._m2l_global_precompute_mis(src_expansion)
+        circulant_matrix_mis, needed_vector_terms, max_mi = \
+            self._m2l_translation_classes_dependent_data_mis(src_expansion)
+
+        circulant_matrix_ident_to_index = dict((ident, i) for i, ident in
+                                enumerate(circulant_matrix_mis))
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Fixed in dfd4d94587787ed10bd4434b30d8b7da8e9f51ae

File: sumpy/expansion/local.py [hunk 36]
```diff
@@ -360,39 +512,37 @@ def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
...
+
+            if self.use_preprocessing_for_m2l:
+                assert m2l_translation_classes_dependent_data is not None
+                assert len(src_coeff_exprs) == len(
+                        m2l_translation_classes_dependent_data)
+                return [a*b for a, b in zip(m2l_translation_classes_dependent_data,
+                    src_coeff_exprs)]
 
-            derivatives_full = [0]*len(toeplitz_matrix_coeffs)
+            derivatives_full = [0]*len(circulant_matrix_mis)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Fixed in bc1e172b90ed5d485d1842e4623f7a7a9d950d94

File: sumpy/expansion/local.py [hunk 37]
```diff
@@ -630,9 +784,92 @@ def evaluate(self, kernel, coeffs, bvec, rscale, sac=None):
...
+                m2l_translation_classes_dependent_data[:2*order], \
+                    m2l_translation_classes_dependent_data[2*order:]
+            first_column_toeplitz = last_column_m2l
+            first_row_toeplitz = list(reversed(first_row_m2l))
+
+            first_column_circulant = list(first_column_toeplitz) + \
+                    list(reversed(first_row_toeplitz))
+            return fft(first_column_circulant, sac)
+
+        return m2l_translation_classes_dependent_data
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/local.py]: Fixed in bc1e172b90ed5d485d1842e4623f7a7a9d950d94

File: sumpy/expansion/local.py (hunk 27 above)
Comment (isuruf) [reply sumpy/expansion/local.py]: Fixed in b5cdb037bd659c9f26961aa7d9f7434ddd4a82d0

File: sumpy/fmm.py (hunk 28 above)
Comment (isuruf) [reply sumpy/fmm.py:331]: Added a FIXME in 66c1296bd6e701fc0d770b8613cef7bbb6c20a21

File: sumpy/tools.py (hunk 29 above)
Comment (isuruf) [reply sumpy/tools.py:925]: Done in 0c59fc4538d1d956b659a748745888800620d785

File: sumpy/e2e.py [hunk 38]
```diff
@@ -148,30 +153,42 @@ class E2EFromCSR(E2EBase):
     default_name = "e2e_from_csr"
 
     def __init__(self, ctx, src_expansion, tgt_expansion,
-            name=None, device=None, use_precomputed_exprs=False):
+            name=None, device=None,
+            m2l_use_translation_classes_dependent_data=False):
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/e2e.py]: Done in 2fec1d09e1c04df1c733dd14f76cc20e0bbd5ba0. Let me know what you think

[the remark this answers is not in the document]
Comment (inducer) [reply sumpy/e2e.py]: Thanks for doing that. Upon looking at it, I think it's about a wash. I'm not loving all the code duplication. On the other hand, I like that there's a code path preserved that doesn't have M2L all throughout. I think it's fine to keep this way.

**Comment by inducer** [comment]
This looks good. Thanks for being patient with me and seeing this through to completion!

**Pull request merged**